### ISSUE TITLE
Cisco 2950 config fetch failed

### ISSUE THREAD

**Issue by polunds** [ytti/oxidized#1460]
Hi, oxidized cant fetch config from cisco 2950 via ssh 
my versions:

```
root@flowcol1:~/.config/oxidized/logs# uname -a
Linux flowcol1 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:39:52 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

```
root@flowcol1:~/.config/oxidized/logs# cat  /etc/debian_version
buster/sid
```

```
oxidized (0.24.0)
oxidized-script (0.5.1)
oxidized-web (0.11.1)


root@flowcol1:~/.config/oxidized# oxidized
Puma starting in single mode...
* Version 3.11.4 (ruby 2.5.1-p57), codename: Love Song
* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://10.168.6.10:8888
Use Ctrl-C to stop

```

config:

```
---
username: oxidized
password: password
model: ios
resolve_dns: false
interval: 3600
use_syslog: true
debug: true
threads: 30
timeout: 20
retries: 0
prompt: !ruby/regexp /^([\w.@-]+[#>]\s?)$/
rest: 10.168.6.10:8888
next_adds_job: false
vars: {}
groups: {}
models: {}
pid: "/root/.config/oxidized/pid"
stats:
  history_size: 10
input:
  default: ssh
  debug: true
  ssh:
    secure: false
  ftp:
    passive: true
  utf8_encoded: true
output:
  default: git
  git:
    user: Oxidized
    email: o@example.com
    repo: "/root/.config/oxidized/default-repo.git"
source:
  default: http
  debug: false
  http:
    secure: false
    url: https://10.168.6.8/api/v0/oxidized
    map:
      name: hostname
      model: os
      group: group
    headers:
      X-Auth-Token: 4bb25184344d4f88a74e2e463124e983
    gpg: false
model_map:
  cisco: ios
  juniper: junos

```

debug logs:

```
Jul 25 09:05:18 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 2 of 159
Jul 25 09:08:13 flowcol1 oxidized[18639]: message repeated 175 times: [ lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 2 of 159]
Jul 25 09:08:13 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: Added 10.168.7.2 to the job queue
Jul 25 09:08:13 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: 1 jobs running in parallel
Jul 25 09:08:13 flowcol1 oxidized[18639]: lib/oxidized/job.rb: Starting fetching process for 10.168.7.2 at 2018-07-25 06:08:13 UTC
Jul 25 09:08:13 flowcol1 oxidized[18639]: AUTH METHODS::["none", "publickey", "password"]
Jul 25 09:08:13 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb: Connecting to 10.168.7.2
Jul 25 09:08:14 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: 1 jobs running in parallel
Jul 25 09:08:14 flowcol1 tac_plus[35062]: 10.168.7.2 shell login for 'oxidized' from 10.168.6.10 on tty1 succeeded
Jul 25 09:08:14 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb: expecting [/^([\w.@()-]+[#>]\s?)$/] at 10.168.7.2
Jul 25 09:08:15 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: 1 jobs running in parallel
Jul 25 09:08:15 flowcol1 oxidized[18639]: lib/oxidized/input/cli.rb: Running post_login commands at 10.168.7.2
Jul 25 09:08:15 flowcol1 oxidized[18639]: lib/oxidized/input/cli.rb: Running post_login command: "terminal length 0", block: nil at 10.168.7.2
Jul 25 09:08:15 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb terminal length 0 @ 10.168.7.2 with expect: /^([\w.@()-]+[#>]\s?)$/
Jul 25 09:08:15 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb: expecting [/^([\w.@()-]+[#>]\s?)$/] at 10.168.7.2
Jul 25 09:08:16 flowcol1 oxidized[18639]: lib/oxidized/input/cli.rb: Running post_login command: "terminal width 0", block: nil at 10.168.7.2
Jul 25 09:08:16 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb terminal width 0 @ 10.168.7.2 with expect: /^([\w.@()-]+[#>]\s?)$/
Jul 25 09:08:16 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb: expecting [/^([\w.@()-]+[#>]\s?)$/] at 10.168.7.2
Jul 25 09:08:16 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: 1 jobs running in parallel
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/model/model.rb Collecting commands' outputs
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/model/model.rb Executing show version
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb show version @ 10.168.7.2 with expect: /^([\w.@()-]+[#>]\s?)$/
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb: expecting [/^([\w.@()-]+[#>]\s?)$/] at 10.168.7.2
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: 1 jobs running in parallel
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/model/model.rb Executing show vtp status
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb show vtp status @ 10.168.7.2 with expect: /^([\w.@()-]+[#>]\s?)$/
Jul 25 09:08:17 flowcol1 oxidized[18639]: lib/oxidized/input/ssh.rb: expecting [/^([\w.@()-]+[#>]\s?)$/] at 10.168.7.2
Jul 25 09:08:18 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: 1 jobs running in parallel
Jul 25 09:08:37 flowcol1 oxidized[18639]: message repeated 19 times: [ lib/oxidized/worker.rb: 1 jobs running in parallel]
Jul 25 09:08:37 flowcol1 oxidized[18639]: 10.168.7.2 raised Timeout::Error with msg "execution expired"
Jul 25 09:08:37 flowcol1 oxidized[18639]: lib/oxidized/node.rb: Oxidized::SSH failed for 10.168.7.2
Jul 25 09:08:37 flowcol1 oxidized[18639]: lib/oxidized/job.rb: Config fetched for 10.168.7.2 at 2018-07-25 06:08:37 UTC
Jul 25 09:08:38 flowcol1 oxidized[18639]: 10.168.7.2 status no_connection, retries exhausted, giving up
Jul 25 09:08:38 flowcol1 oxidized[18639]: lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 3 of 159

```

ssh log:
sometime oxidized fetch show version

```
root@flowcol1:~/.config/oxidized/logs# cat 10.168.7.2-ssh

ca.access.K1-1-4.ZAL.c2950#terminal length 0
ca.access.K1-1-4.ZAL.c2950#terminal width 0
ca.access.K1-1-4.ZAL.c2950#show version
Cisco Internetwork Operating System Software
IOS (tm) C2950 Software (C2950-I6K2L2Q4-M), Version 12.1(22)EA13, RELEASE SOFTWARE (fc2)
Technical Support: http://www.cisco.com/techsupport
Copyright (c) 1986-2009 by cisco Systems, Inc.
Compiled Fri 27-Feb-09 22:20 by amvarma
Image text-base: 0x80010000, data-base: 0x80680000

ROM: Bootstrap program is C2950 boot loader

ca.access.K1-1-4.ZAL.c2950 uptime is 5 weeks, 22 hours, 36 minutes
System returned to ROM by power-on
System image file is "flash:/c2950-i6k2l2q4-mz.121-22.EA13.bin"


This product contains cryptographic features and is subject to United
States and local country laws governing import, export, transfer and
use. Delivery of Cisco cryptographic products does not imply
third-party authority to import, export, distribute or use encryption.
Importers, exporters, distributors and users are responsible for
compliance with U.S. and local country laws. By using this product you
agree to comply with applicable laws and regulations. If you are unable
to comply with U.S. and local laws, return this product immediately.

A summary of U.S. laws governing Cisco cryptographic products may be found at:
http://www.cisco.com/wwl/export/crypto/tool/stqrg.html

If you require further assistance please contact us by sending email to
export@cisco.com.

cisco WS-C2950-12 (RC32300) processor (revision R0) with 19912K bytes of memory.
Processor board ID FCZ1032X0A6
Last reset from system-reset
Running Standard Image
12 FastEthernet/IEEE 802.3 interface(s)

32K bytes of flash-simulated non-volatile configuration memory.
Base ethernet MAC Address: 00:18:B9:D5:44:40
Motherboard assembly number: 73-5782-13
Power supply part number: 34-0965-01
Motherboard serial number: FOC10280CU5
Power supply serial number: DTH10171C5R
Model revision number: R0
Motherboard revision number: A0
Model number: WS-C2950-12
System serial number: FCZ1032X0A6
Configuration register is 0xF

ca.access.K1-1-4.ZAL.c2950#

```
some time it just 

```
root@flowcol1:~/.config/oxidized/logs# cat 10.168.7.2-ssh

ca.access.K1-1-4.ZAL.c2950#terminal length 0
ca.access.K1-1-4.ZAL.c2950#
```

what can i do to fix this?

**Comment by fatman00** [comment]
To me it looks like the switch terminates the SSH session. Have you tried running the same commands manually through putty a couple of times and see if the same error occurs?

I have seen some devices just cant handle a lot of show commands in quick succession because of memory leaks.

**Comment by nickhilliard** [comment]
you need to upgrade:  this is an ios software bug. it was definitely still present in 12.1(22)EA14 and fixed by 12.1(22)EA8a, probably sooner.

**Comment by nickhilliard** [comment]
wrong way around.  Downgrade to EA10a or lower.
